File: 40.combination.sum.II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort();
    ret = [];
    function dfs(i, sum, l){
        if (sum == 0){
            ret.push(l.slice());
            return;
        }
        if (i <0 || sum < 0) return;
        low = i-1;
        while (low >=0  && candidates[low] == candidates[i]) low--;
        dfs(low, sum ,l);
        l.push(candidates[i]);
        dfs(i-1, sum-candidates[i], l);
        l.pop();
    }
    dfs(candidates.length-1, target, []);
    return ret;
};

File: 40. Combination Sum II.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        n = len(candidates)
        q = [(0, 0, [])]
        ret = []
        while len(q) > 0:
            i, s, nums = q.pop()
            #print(i, s, nums)
            if s == target:
                ret.append(nums)
                continue
            if i >= n or s + candidates[i] > target:
                continue
            k = i + 1
            while k < n and candidates[i] == candidates[k]:
                k += 1
            q.append((k, s, nums))
            j = i + 1
            while j <= k:
                if s + candidates[i] * (j-i) <= target:
                    q.append((k, s + candidates[i] * (j-i), nums+[candidates[i]]*(j-i)))
                else:
                    break
                j += 1
        return ret
